src/styles/bannerCommon.ts

```diff
@@ -1,240 +1,336 @@
-        padding: ${space[4]}px;
-        min-height: ${imgHeight}px;
-        ${from.desktop} {
-            width: 60%;
-        }
-    `,
-
-    bottomRightComponent: css`
-        display: flex;
-        justify-content: center;
-        width: 100%;
-        max-height: 100%;
-
-        ${from.desktop} {
-            align-self: flex-end;
-            padding-right: ${space[4]}px;
-            max-width: 45%;
-            justify-content: flex-end;
-        }
-
-        ${from.leftCol} {
-            padding-right: 0;
-            justify-content: space-between;
+export interface StyleData {
+    styleBackground: string;
+    styleHeader: string;
+    styleBody: string;
+    styleHighlight: string;
+    styleHighlightBackground: string;
+    styleButton: string;
+    styleButtonBackground: string;
+    styleButtonHover: string;
+    styleClose: string;
+    styleCloseBackground: string;
+    styleCloseHover: string;
+}
+
+const colorStringStyles = [
+    'styleBackground',
+    'styleHeader',
+    'styleBody',
+    'styleHighlight',
+    'styleHighlightBackground',
+    'styleButton',
+    'styleButtonBackground',
+    'styleButtonHover',
+    'styleClose',
+    'styleCloseBackground',
+    'styleCloseHover',
+];
+
+type Styles = keyof StyleData;
+
+export const selfServeStyles = (userVals: Extras, defs: StyleData) => {
+    const style: StyleData = Object.assign({}, defs);
+    const defKeys: Styles[] = Object.keys(defs) as Styles[];
+
+    const canvas = document.createElement('canvas');
+    canvas.width = 1;
+    canvas.height = 1;
+    const ctx = canvas.getContext('2d', { willReadFrequently: true });
+
+    defKeys.forEach((key) => {
+        const userVal = userVals[key];
+        const defVal = defs[key];
+
+        // If user val is undefined, or an empty string, use default val
+        if (userVal == null || !userVal) {
+            style[key] = defVal;
+        } else {
+            let flag = true;
+
+            // Protect against CSS injection
+            const item = userVal.split(';')[0];
+            if (item == null || !item) {
+                flag = false;
+            }
+
+            // The following code is some very naive checks for legitimate CSS color strings
+
+            // Protect against unequal number of open/close parentheses
+            else if (item.includes('(')) {
+                const opens = item.split('').filter((c) => c === '(');
+                const closes = item.split('').filter((c) => c === ')');
+                if (opens.length !== closes.length) {
+                    flag = false;
+                }
+            }
+            // Check for legitimate CSS color string values
+            if (flag && ctx && colorStringStyles.includes(key)) {
+                ctx.fillStyle = 'transparent';
+                ctx.fillStyle = item;
+                ctx.clearRect(0, 0, 1, 1);
+                ctx.fillRect(0, 0, 1, 1);
+                const iData = ctx.getImageData(0, 0, 1, 1);
```

Comment:
would `getImageData` fail if the colour was invalid?

Reply:
`getImageData` will always return an imageData object. If the colour is invalid then the returned pixel will be transparent black `[0, 0, 0, 0]` which is how we can identify the user supplied colour string is malformed

Reply:
ok, please can you add this in a comment so it's clear to people like me

Reply:
Done